### PR TITLE
ENH: add doctests, support for functionals and nonlinear ops

### PR DESCRIPTION
**Pull request by kohr-h**

### CODE REVIEW

File: odl/contrib/theano/layer.py [hunk 1]
```diff
@@ -17,60 +17,260 @@
...
-    """Wraps ODL operator to use with theano.
+
+    """Wrap an ODL operator as a Theano operator.
 
     The operator works with theanos automatic differentiation if the operator
     implements `Operator.derivative` and `Operator.adjoint`.
     """
 
-    # Properties used by theano to check for equality
+    # Properties used by theano for __eq__ and __hash__
```
Comment (adler-j) [inline odl/contrib/theano/layer.py]: Also for `__repr__`

File: odl/contrib/theano/layer.py [hunk 2]
```diff
@@ -17,60 +17,260 @@
...
+        >>> matrix_op.operator is op
+        True
+
+        Create a functional, i.e., an operator with scalar output:
+
+        >>> space = odl.rn(3)
+        >>> functional = odl.solvers.L2NormSquared(space)
+        >>> func_op = TheanoOperator(functional)
+        >>> x = theano.tensor.dvector()
+        >>> apply = func_op.make_node(x)
```
Comment (adler-j) [inline odl/contrib/theano/layer.py:64]: func_op(x) does not work?

File: odl/contrib/theano/layer.py [hunk 3]
```diff
@@ -17,60 +17,260 @@
...
+        Perform a matrix multiplication:
+
+        >>> space = odl.rn(3)
+        >>> matrix = np.array([[1, 0, 1],
+        ...                    [0, 1, 1]], dtype=float)
+        >>> op = odl.MatrixOperator(matrix, domain=space)
+        >>> matrix_op = TheanoOperator(op)
+        >>> x = theano.tensor.dvector('x')
+        >>> op_x = matrix_op(x)
+        >>> op_func = theano.function([x], [op_x])
```
Comment (adler-j) [inline odl/contrib/theano/layer.py]: Don't think you need parenthisis around `op_x`, removing should improve the print below

File: odl/contrib/theano/layer.py [hunk 4]
```diff
@@ -17,60 +17,260 @@
...
+        Examples
+        --------
+        Perform a matrix multiplication:
+
+        >>> space = odl.rn(3)
+        >>> matrix = np.array([[1, 0, 1],
+        ...                    [0, 1, 1]], dtype=float)
+        >>> op = odl.MatrixOperator(matrix, domain=space)
+        >>> matrix_op = TheanoOperator(op)
+        >>> x = theano.tensor.dvector('x')
```
Comment (adler-j) [inline odl/contrib/theano/layer.py]: Don't we need to import theano here?

Comment (adler-j) [reply odl/contrib/theano/layer.py]: Also screw naming them in doctests

File: odl/contrib/theano/layer.py [hunk 5]
```diff
@@ -83,12 +283,48 @@ def grad(self, inputs, output_grads):
...
+                """Create a node for the computation graph."""
+                x = theano.tensor.as_tensor_variable(x)
+                v = theano.tensor.as_tensor_variable(v)
+                return theano.Apply(self, [x, v], [x.type()])
+
+            def perform(self, node, inputs_storage, output_storage):
+                """Evaluate this node's computation.
+
+                This method computes ::
+
```
Comment (adler-j) [inline odl/contrib/theano/layer.py:310]: Remove blankline

File: odl/contrib/theano/layer.py [hunk 6]
```diff
@@ -83,12 +283,48 @@ def grad(self, inputs, output_grads):
...
+                if scale != 1.0:
+                    out[0] *= scale
+
+            def infer_shape(self, node, input_shapes):
+                """Return a list of output shapes based on ``input_shapes``."""
+                # Need to convert to native to avoid error in theano from
+                # future.int
+                return [tuple(native(si) for si in op.operator.domain.shape)]
+
+        r_op = TheanoJacobian()
```
Comment (adler-j) [inline odl/contrib/theano/layer.py]: Now that you actually fix this, you should also remove the "ignore unused variables" in the code (test and example)

**Review by adler-j — APPROVED** [review]
Excellent work (almost feels like you had this laying around?)

some minor stuff

**Comment by kohr-h** [comment]
> Excellent work (almost feels like you had this laying around?)

Yes, as I wrote [here](https://github.com/odlgroup/odl/issues/1098#issuecomment-325156315) 😉 

Anyway, I'll make the amendments.

File: odl/contrib/theano/layer.py (hunk 1 above)
Comment (kohr-h) [reply odl/contrib/theano/layer.py]: Good point

File: odl/contrib/theano/layer.py (hunk 2 above)
Comment (kohr-h) [reply odl/contrib/theano/layer.py:64]: It should, doctest seems incomplete regarding that. Probably I just wanted to see the "scalar" word in the output.

File: odl/contrib/theano/layer.py (hunk 4 above)
Comment (kohr-h) [reply odl/contrib/theano/layer.py]: Will check

File: odl/contrib/theano/layer.py [hunk 7]
```diff
@@ -17,60 +17,260 @@
...
+        ...                    [0, 1, 1]], dtype=float)
+        >>> op = odl.MatrixOperator(matrix, domain=space)
+        >>> matrix_op = TheanoOperator(op)
+        >>> x = theano.tensor.dvector('x')
+        >>> op_x = matrix_op(x)
+        >>> op_func = theano.function([x], [op_x])
+        >>> op_func([1, 2, 3])
+        [array([ 4.,  5.])]
+
+        Evaluate a unctional, i.e., an operator with scalar output:
```
Comment (kohr-h) [inline odl/contrib/theano/layer.py]: Oops, typo

File: odl/contrib/theano/layer.py (hunk 2 above)
Comment (kohr-h) [reply odl/contrib/theano/layer.py:64]: The result is `TheanoOperator{operator=L2NormSquared(rn(3))}.0`, which is not bad, but I'm not sure if it's good enough for a doctest.

File: odl/contrib/theano/layer.py (hunk 3 above)
Comment (kohr-h) [reply odl/contrib/theano/layer.py]: Not quite sure what you want to say, but maybe the answer is this: ~~`theano.function` always needs a list of inputs and a list (or dictionary) of outputs. I could add `[0]` but that wouldn't make it look more natural.~~ Seems like theano's documentation just doesn't mention that it works with single output, too. I'll change that.

File: odl/contrib/theano/layer.py (hunk 5 above)
Comment (kohr-h) [reply odl/contrib/theano/layer.py:310]: That would ruin the typesetting, keeping it as-is.

File: odl/contrib/theano/layer.py (hunk 3 above)
Comment (adler-j) [reply odl/contrib/theano/layer.py]: Quite sure you can do `op_func = theano.function([x], op_x)`

as is done here: https://github.com/adler-j/odl/blob/2888f9ba927cf67e21b50c7fa75c703802b299a3/odl/contrib/theano/examples/theano_layer_matrix.py#L34

**Comment by kohr-h** [comment]
Done everything, test and example now compute the gradient of a functional. Fixed some minor stuff, too.

File: odl/contrib/theano/examples/theano_layer_matrix.py [hunk 8]
```diff
@@ -4,56 +4,56 @@
...
+We also demonstrate that we can compute the gradient of the scalar-valued
+squared L2-norm function properly using either Theano or ODL.
 """
 
 import theano
 import theano.tensor as T
 import numpy as np
 import odl
-import odl.contrib.theano
+from odl.contrib import theano as theano_wrapper
```
Comment (kohr-h) [inline odl/contrib/theano/examples/theano_layer_matrix.py]: Maybe I shouldn't have done this

File: odl/util/testutils.py [hunk 9]
```diff
@@ -660,10 +660,17 @@ def test(arguments=None):
     pytest.main(args)
 
 
-def run_doctests(skip_if=False):
+def run_doctests(skip_if=False, extraglobs=None):
```
Comment (kohr-h) [inline odl/util/testutils.py]: I found this useful, but probably I should break it out from this pure contrib stuff.

Comment (adler-j) [reply odl/util/testutils.py]: Agree, break it out.

**Comment by kohr-h** [comment]
I removed the extraglobs stuff again, at least for me it just works without any changes (pytest and running the module directly).
I didn't bother keeping the history clean since you'll squash everything into one commit anyway, right @adler-j? 
So from my perspective this can go in.

**Comment by adler-j** [comment]
Since its a PR into a PR I'll fix any minor stuff later

**Pull request merged**